src/functions.cpp

```diff
@@ -0,0 +1,169 @@
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  return add(origin, {}, name, return_type, params);
+}
+
+const Function &FunctionRegistry::add(Function::Origin origin,
+                                      std::string_view ns,
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  all_funcs_.push_back(std::make_unique<Function>(
+      origin, std::string{ name }, return_type, params));
+  Function &new_func = *all_funcs_.back().get();
+
+  FqName fq_name{
+    .ns = std::string{ ns },
+    .name = std::string{ name },
+  };
+
+  // Check for duplicate function definitions
+  // The assumption is that builtin functions are all added to the registry
+  // before any user-defined functions.
+  // Builtin functions can be duplicated. Other functions can not.
+  for (const Function &func : funcs_by_fq_name_[fq_name]) {
+    if (func.origin() != Function::Origin::Builtin) {
+      throw std::runtime_error("Cannot redefine existing function: " +
+                               fq_name.str() + param_types_str(params));
+    }
+  }
+
+  funcs_by_fq_name_[fq_name].push_back(new_func);
+  return new_func;
+}
+
+namespace {
+bool can_implicit_cast(const SizedType &from, const SizedType &to)
+{
+  if (from.FitsInto(to))
+    return true;
+
+  if (from.IsStringTy() && to.IsPtrTy() && to.GetPointeeTy()->IsIntTy() &&
+      to.GetPointeeTy()->GetSize() == 1) {
+    // Allow casting from string to int8* or uint8*
+    return true;
+  }
+
+  // Builtin and script functions do not care about string sizes. External
+  // functions cannot be defined to accept string types (they'd take char*)
+  if (from.IsStringTy() && to.IsStringTy())
+    return true;
+
+  return false;
+}
+} // namespace
+
+/*
+ * Find the best function called 'name' for the given argument types.
+ *
+ * Returns either a single function or nullptr, when no such function exists.
+ *
+ * When there are multiple candidate functions with the same name, prefer the
+ * non-builtin over the builtin function.
+ *
+ * Valid functions have the correct name and all arguments can be implicitly
+ * casted into all parameter types.
+ */
+const Function *FunctionRegistry::get(std::string_view ns,
+                                      std::string_view name,
+                                      const std::vector<SizedType> &arg_types,
+                                      std::optional<location> loc) const
+{
+  FqName fq_name = {
+    .ns = std::string{ ns },
+    .name = std::string{ name },
+  };
+  auto it = funcs_by_fq_name_.find(fq_name);
+  if (it == funcs_by_fq_name_.end()) {
+    LOG(ERROR, out_) << "Function not found: '" << name << "'";
```

Comment:
Maybe a silly question but a user won't actually get here right? I assume we'd surface an error in one of the passes with a bit of underlined code about the not found function. Long winded way of asking if this should be `LOG(BUG)`.

Reply:
The intention is that this code gives the error for functions that don't exist. It'd avoid duplication of error messages if there are multiple callers of `FunctionRegistry::get`, and `FunctionRegistry` also has some more information that it can use to give more useful error messages in the future, e.g. something like:
```
ERROR: Function not found: "prin"
HINT: Did you mean: "print"?
```

I did miss out the `loc` argument to `LOG` though, which I've added now.